RFC: Add some functionality from winedbg to enable loading of debug informations for PE images

bernhardu opened the pull request:
Since wine switched to use mostly PE images for their dlls the replay gdb is
not able ro load the symbol informations for these anymore.

This patches do basically the same as 'winedbg --gdb' does, mostly tell gdb to ask rr
to get the library list instead of inspecting the linux dynamic loader by himself.

Currently recording should behave exactly the same as without these patches.
One note: the whole wine process tree needs to be in the record, or something like in 2805 can happen.

My replayings were done by 'rr replay --target-wine -p <PID> -g <time>' to replay
the right process (list with 'rr ps') and testing some values for -g to land when wine
finished process creating. This is kind of try and error currently.

I tested some combinations of 32- and 64-bit rr and wine builds.
One thing I found is, a 'finish' in the __wine_syscall_dispatcher frame runs until the end of the recording.
But can be worked around by single stepping some instructions out with 'stepi'.

What would be a proper place for the structs IMAGE_DOS_HEADER and related defines?

Previous discussion in #2751.
What do you think?

bernhardu committed:
Add library-list to gdb interface.

This just iterates through the memory mapped files,
first searching for PE signatures, and for regular
shared libraries just for executable mappings.

Related to #2751.

bernhardu committed:
Add option --target-wine

This convinces gdb to load the shared objects list
through the gdb interface instead of accessing memory directly.

Related to #2751.

bernhardu committed:
Notify gdb about changed executable memory mappings.

This is done by intercepting mprotect and unmap executions.
Currently not completely reliable.
In case of doubt issue 'nosharedlibrary' and 'sharedlibrary'
to have gdb reload all.

Related to #2751.

rocallahan commented on src/GdbServer.cc:
```diff
@@ -1771,7 +1774,14 @@ void GdbServer::launch_gdb(ScopedFd& params_pipe_fd,
     push_target_remote_cmd(args, string(params.host), params.port);
   }
-  args.push_back(params.exe_image);
+  if (!target_wine) {
```
Flip the true/false block order to avoid ! here.

bernhardu replied:
Will do.

rocallahan commented on src/GdbConnection.cc:
```diff
@@ -423,17 +423,17 @@ static string read_target_desc(const char* file_name) {
       return "amd64-linux.xml";
     case GdbConnection::CPU_AVX:
-      return "i386-avx-linux.xml";
+      return target_wine ? "i386-avx-wine.xml" : "i386-avx-linux.xml";
```
Why do we need special Wine support here?

bernhardu replied:
That is how winedbg works currently. It supplies a "custom" xml file, and later executes the "sharedlibrary" command.
If I supply gdb with the regular xml file it uses the "svr4" ops and then it does not "qXfer:libraries" anymore.

That is the backtrace where it requests libraries "winedbg way":
```
(rr) bt
#0  remote_target::remote_read_qxfer (this=0x55f6dbcc09d0, object_name=0x55f6d9ceed86 "libraries", annex=0x0, readbuf=0x55f6dbd3dd60 "`\263>\024+\177", offset=0, len=4096, xfered_len=0x7fff4af67e48, packet=0x55f6d9f5d738 <remote_protocol_packets+504>) at /build/gdb-Nav6Es/gdb-10.1/gdb/remote.c:10881
#1  0x000055f6d9acb1be in target_xfer_partial (ops=0x55f6dbcc09d0, object=TARGET_OBJECT_LIBRARIES, annex=0x0, readbuf=0x55f6dbd3dd60 "`\263>\024+\177", writebuf=0x0, offset=0, len=<optimized out>, xfered_len=0x7fff4af67e48) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1160
#2  0x000055f6d9acda0f in target_read_partial (xfered_len=0x7fff4af67e48, len=4096, offset=0, buf=<optimized out>, annex=0x0, object=TARGET_OBJECT_LIBRARIES, ops=0x55f6dbcc09d0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1387
#3  target_read_alloc_1<char> (ops=0x55f6dbcc09d0, object=object@entry=TARGET_OBJECT_LIBRARIES, annex=annex@entry=0x0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1722
#4  0x000055f6d9accf90 in target_read_stralloc (ops=<optimized out>, object=object@entry=TARGET_OBJECT_LIBRARIES, annex=annex@entry=0x0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1761
#5  0x000055f6d9a75f57 in solib_target_current_sos () at /build/gdb-Nav6Es/gdb-10.1/gdb/solib-target.c:237
#6  0x000055f6d9a7891c in update_solib_list (from_tty=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib.c:727
#7  0x000055f6d9a78f07 in solib_add (pattern=0x0, from_tty=1, readsyms=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib.c:961
#8  0x000055f6d985f76c in cmd_func (cmd=0x55f6dbc78e50, args=0x0, from_tty=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/cli/cli-decode.c:2181
#9  0x000055f6d9ad7a54 in execute_command (p=<optimized out>, p@entry=<error reading variable: value has been optimized out>, from_tty=1, from_tty@entry=<error reading variable: value has been optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/top.c:668
#10 0x000055f6d99af833 in catch_command_errors (command=<optimized out>, arg=<optimized out>, from_tty=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:457
#11 0x000055f6d99b3147 in captured_main_1 (context=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1218
#12 0x000055f6d99b357b in captured_main (data=data@entry=0x7fff4af68420) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1243
#13 gdb_main (args=args@entry=0x7fff4af68450) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1268
#14 0x000055f6d97cd90c in main (argc=<optimized out>, argv=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/gdb.c:32
(rr) print cmd.name
$1 = 0x55f6d9d17b49 "sharedlibrary"
```


But while collecting this I found following interesting:
```
...remote_target::remote_read_qxfer (..., object_name=0x55f6d9ced563 "memory-map", ...
```
Maybe implementing this could improve situation.


And second, it doesn't "qXfer:libraries" but it looks like it does "qXfer:libraries-svr4" instead:
```
(rr) bt
#0  remote_target::remote_read_qxfer (this=0x562bbc3109d0, object_name=0x562bba9c7af9 "libraries-svr4", annex=0x0, readbuf=0x562bbc38d980 "", offset=0, len=4096, xfered_len=0x7fffe2898ca8, packet=0x562bbac18750 <remote_protocol_packets+528>) at /build/gdb-Nav6Es/gdb-10.1/gdb/remote.c:10881
#1  0x0000562bba7861be in target_xfer_partial (ops=0x562bbc3109d0, object=TARGET_OBJECT_LIBRARIES_SVR4, annex=0x0, readbuf=0x562bbc38d980 "", writebuf=0x0, offset=0, len=<optimized out>, xfered_len=0x7fffe2898ca8) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1160
#2  0x0000562bba788a0f in target_read_partial (xfered_len=0x7fffe2898ca8, len=4096, offset=0, buf=<optimized out>, annex=0x0, object=TARGET_OBJECT_LIBRARIES_SVR4, ops=0x562bbc3109d0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1387
#3  target_read_alloc_1<char> (ops=0x562bbc3109d0, object=object@entry=TARGET_OBJECT_LIBRARIES_SVR4, annex=annex@entry=0x0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1722
#4  0x0000562bba787f90 in target_read_stralloc (ops=<optimized out>, object=object@entry=TARGET_OBJECT_LIBRARIES_SVR4, annex=annex@entry=0x0) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:1761
#5  0x0000562bba72da95 in svr4_current_sos_via_xfer_libraries (list=0x7fffe2898dd0, annex=0x0) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib-svr4.c:1235
#6  0x0000562bba72f0e6 in svr4_current_sos_direct (info=0x562bbc30d590) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib-svr4.c:1381
#7  0x0000562bba72f7a5 in svr4_current_sos_1 (info=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib-svr4.c:1446
#8  svr4_current_sos () at /build/gdb-Nav6Es/gdb-10.1/gdb/solib-svr4.c:1455
#9  0x0000562bba73391c in update_solib_list (from_tty=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib.c:727
#10 0x0000562bba733f07 in solib_add (pattern=0x0, from_tty=1, readsyms=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/solib.c:961
#11 0x0000562bba51a76c in cmd_func (cmd=0x562bbc2c8e50, args=0x0, from_tty=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/cli/cli-decode.c:2181
#12 0x0000562bba792a54 in execute_command (p=<optimized out>, p@entry=<error reading variable: value has been optimized out>, from_tty=1, from_tty@entry=<error reading variable: value has been optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/top.c:668
#13 0x0000562bba66a833 in catch_command_errors (command=<optimized out>, arg=<optimized out>, from_tty=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:457
#14 0x0000562bba66e147 in captured_main_1 (context=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1218
#15 0x0000562bba66e57b in captured_main (data=data@entry=0x7fffe28992f0) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1243
#16 gdb_main (args=args@entry=0x7fffe2899320) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1268
#17 0x0000562bba48890c in main (argc=<optimized out>, argv=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/gdb.c:32
(rr) print cmd.name
$2 = 0x562bba9d2b49 "sharedlibrary"
```


I will investigate what those two options reveal ...

bernhardu replied:
"memory-map" seems more to describe which memory is used, but not related to libraries. So that is an dead end.

bernhardu replied:
Just some more context (at least for me, to make it easier to remember):
Gdb does by default use the `solib_target_so_ops`.
Just when all received target definitions match then gdb starts to use `svr4_so_ops`.
Therefore winedbg seems to not supply a perfect definition on purpose.

That verification takes place here:
```
(rr) bt
#0  0x0000560d3ea8b84c in tdesc_find_register_early (name=name@entry=0x560d3ec4c467 "orig_rax", feature=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/target-descriptions.c:806
#1  tdesc_numbered_register (feature=<optimized out>, data=data@entry=0x560d3f1d2f40, regno=regno@entry=154, name=name@entry=0x560d3ec4c467 "orig_rax") at /build/gdb-Nav6Es/gdb-10.1/gdb/target-descriptions.c:819
#2  0x0000560d3e7d9e3e in amd64_linux_init_abi (info=..., gdbarch=0x560d3f1df530) at /build/gdb-Nav6Es/gdb-10.1/gdb/amd64-linux-tdep.c:1878
#3  0x0000560d3e92fdc2 in i386_gdbarch_init (info=..., arches=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/i386-tdep.c:8705
#4  0x0000560d3e90ed52 in gdbarch_find_by_info (info=...) at /build/gdb-Nav6Es/gdb-10.1/gdb/gdbarch.c:5496
#5  0x0000560d3e7e010e in gdbarch_update_p (info=...) at /build/gdb-Nav6Es/gdb-10.1/gdb/arch-utils.c:538
#6  0x0000560d3ea8aa39 in target_find_description () at /build/gdb-Nav6Es/gdb-10.1/gdb/target-descriptions.c:569
#7  target_find_description () at /build/gdb-Nav6Es/gdb-10.1/gdb/target-descriptions.c:530
#8  0x0000560d3ea39ac3 in remote_target::start_remote (this=this@entry=0x560d3f1709d0, from_tty=from_tty@entry=1, extended_p=extended_p@entry=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/remote.c:4644
#9  0x0000560d3ea3a411 in remote_target::open_1 (name=<optimized out>, from_tty=1, extended_p=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/remote.c:5668
#10 0x0000560d3eaa8cbe in open_target (args=0x7ffcd1a0df53 "127.0.0.1:50868", from_tty=1, command=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/target.c:242
#11 0x0000560d3e83b76c in cmd_func (cmd=0x560d3f1115c0, args=0x7ffcd1a0df53 "127.0.0.1:50868", from_tty=1) at /build/gdb-Nav6Es/gdb-10.1/gdb/cli/cli-decode.c:2181
#12 0x0000560d3eab3a54 in execute_command (p=<optimized out>, p@entry=<error reading variable: value has been optimized out>, from_tty=1, from_tty@entry=<error reading variable: value has been optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/top.c:668
#13 0x0000560d3e98b833 in catch_command_errors (command=<optimized out>, arg=<optimized out>, from_tty=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:457
#14 0x0000560d3e98f147 in captured_main_1 (context=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1218
#15 0x0000560d3e98f57b in captured_main (data=data@entry=0x7ffcd1a0c840) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1243
#16 gdb_main (args=args@entry=0x7ffcd1a0c870) at /build/gdb-Nav6Es/gdb-10.1/gdb/main.c:1268
#17 0x0000560d3e7a990c in main (argc=<optimized out>, argv=<optimized out>) at /build/gdb-Nav6Es/gdb-10.1/gdb/gdb.c:32
```

bernhardu replied:
As far as I see libraries-svr4 request depends on having some dynamic loader structures, which do not exist.
Therefore the libraries request seems better suiting for this? ( [qXfer:libraries:read](https://sourceware.org/gdb/current/onlinedocs/gdb/Library-List-Format.html) and [qXfer:libraries-svr4:read](https://sourceware.org/gdb/current/onlinedocs/gdb/Library-List-Format-for-SVR4-Targets.html) )

rocallahan commented on src/GdbConnection.cc:
```diff
@@ -679,14 +679,16 @@ bool GdbConnection::query(char* payload) {
                  ";vContSupported+";
+    if (!features().target_wine) {
+      supported << ";qXfer:exec-file:read+";
+    }
```
Why are we doing this?

bernhardu replied:
That is also like winedbg handles it and while developing this I thought this is needed to work.
But I tested again with the current state and I can present this feature to gdb and it still works. I will remove this.

rocallahan commented on src/GdbConnection.cc:
```diff
@@ -1712,6 +1728,172 @@ void GdbConnection::reply_get_thread_list(const vector<GdbThreadId>& threads) {
+          libraries[name].has_exec = true;
+        }
+      }
+    }
```
I think we really need to search ld.so's library list if we're going to support this protocol feature. Using the mapping list only is probably not good enough.

bernhardu replied:
For now it could be just announced if target_wine is set?

rocallahan replied:
Yes, I think that's a good idea.

bernhardu replied:
Just for the record, I did not see gdb reqeuest "qXfer:libraries" even when announced when in svr4 mode, just a "qXfer:libraries-svr4".

rocallahan commented on src/GdbConnection.cc:
```diff
@@ -1712,6 +1728,172 @@ void GdbConnection::reply_get_thread_list(const vector<GdbThreadId>& threads) {
   consume_request();
 }
 
+/* see wine include/winnt.h*/
```
Let's put the Wine-specific code into its own file, say `src/Wine.cc`.

rocallahan commented on src/AddressSpace.cc:
```diff
@@ -1260,6 +1264,18 @@ void AddressSpace::unmap(Task* t, remote_ptr<void> addr, ssize_t num_bytes) {
+    {
+      if (!executable_mappings_changed_ && km.prot() & PROT_EXEC) {
+        executable_mappings_changed_ = true;
+      }
```
If we use ld.so's library list then I think we may need to set our own breakpoints in the ld.so debug hooks, which makes this a lot more work. But I think it's the right thing to do.

bernhardu replied:
I have to check, but I guess wine is not notifying ld.so when loading PE modules,
so there would be no benefit for target_wine?

rocallahan replied:
Ok.

bernhardu replied:
If the dll is a (with current wine less used) .so file then it
would be done by dynamic loader via dlopen_unix_dll / dlopen.
But I have not found something related to dynamic loader for native dlls.

This is an attempt to outline how winedbg gets notified about a dll loading:
```
ntdll:LdrLoadDll
ntdll:load_dll
ntdll:load_native_dll
ntdll:build_module - SERVER_START_REQ( load_dll )
server:DECL_HANDLER(load_dll)
server:generate_debug_event( current, DbgLoadDllStateChange, dll );
server:alloc_debug_event
server:fill_load_dll_event
winedbg:event_data_to_state_change
winedbg:NtWaitForDebugEvent - SERVER_START_REQ( wait_debug_event )
winedbg:DbgUiWaitStateChange
winedbg:WaitForDebugEvent
winedbg:handle_debug_event
winedbg:LOAD_DLL_DEBUG_EVENT
winedbg:dbg_load_module
dbghelp:SymLoadModuleExW
```
Unfortunately I do not see a hook that would be more suitable or to get the address to do something like `add_breakpoint(p, BKPT_INTERNAL)`.

rocallahan commented:
I think this makes sense but it's going to be a lot of work to get right and potentially fragile if the ld.so interface evolves. (However, we did implement something similar for Pernosco's gdbserver.)

rocallahan commented:
I think it makes sense to only provide a library list in wine-target mode and use the mapped binary list to do it, like what you're doing now.

bernhardu commented:
I made these changes in this iteration:

"Add library-list to gdb interface."
* Moved wine specifics into Wine.cc
* Announce qXfer:libraries just in target-wine mode
* Changed a few comments in the mapping iteration to improve readability.

"Add option --target-wine"
* Remove exclusion of qXfer:exec-file in target-wine mode
* Flipped the true/false block order to avoid !

"Notify gdb about changed executable memory mappings."
* (unchanged)

rocallahan commented on src/Wine.cc:
```diff
@@ -0,0 +1,175 @@
+  char unused1[0xc];
+  uint32_t VirtualAddress;
+  char unused2[0x18];
+};
```
What's the copyright status of this code?

bernhardu replied:
Lines 18, 19, 20, 21, 23, 25, 48, 49, 50 are copied and adapted from [wine's winnt.h](https://source.winehq.org/git/wine.git/blob/HEAD:/include/winnt.h), which is licensed as LGPL as far as I see.
The remaining of the structure definitions I wrote myself by looking at the above file and the output of `wine winedbg --gdb notepad` with `Wine-gdb> ptype /o IMAGE_DOS_HEADER`.
Is this a problem with RRs license?
Would [mingw headers](https://sourceforge.net/p/mingw-w64/mingw-w64/ci/master/tree/mingw-w64-headers/include/winnt.h#l6134) with ZPL license be possible?
Or [boost headers](https://github.com/boostorg/dll/blob/develop/include/boost/dll/detail/pe_info.hpp#L42)?
Is it sufficient to remove the defines and rewrite the other lines?

rocallahan replied:
We don't want LGPL code in rr. The Boost license would be fine, but importing that Boost file as-is is probably unnecessarily complicated.

I suggest just rewriting the code from scratch, yourself, referring to the Boost file as needed, and mention in comments that you did that and that the Boost file is under the Boost (MIT-ish) license.

Keno replied:
Microsoft has an MIT licensed version of these struct defines here: https://github.com/microsoft/xlang/blob/b950b15d53cab2dddc78c65e783ef43042a2e8c9/src/library/impl/meta_reader/pe.h

rocallahan commented on src/Wine.cc:
```diff
@@ -0,0 +1,175 @@
+    if (km.fsname().substr(0, 7) == "/memfd:")
+      continue;
+    if (km.is_vdso() || km.is_heap() || km.is_stack() || km.is_vvar() || km.is_vsyscall())
+      continue;
```
{} around if bodies

rocallahan commented on src/Wine.cc:
```diff
@@ -0,0 +1,175 @@
+    bool ok = true;
+    t->read_bytes_helper(km.start(), sizeof(buf), buf, &ok);
+    if (ok) {
+      struct IMAGE_DOS_HEADER* dosheader = (struct IMAGE_DOS_HEADER*)buf;
```
reinterpret_cast<>

rocallahan commented on src/Wine.cc:
```diff
@@ -0,0 +1,175 @@
+    t->read_bytes_helper(km.start(), sizeof(buf), buf, &ok);
+    if (ok) {
+      struct IMAGE_DOS_HEADER* dosheader = (struct IMAGE_DOS_HEADER*)buf;
+      struct IMAGE_NT_HEADERS* ntheader = (struct IMAGE_NT_HEADERS*)(buf + dosheader->e_lfanew);
```
Technically this is undefined behavior. Don't construct the pointer until you've tested that there is room for `IMAGE_NT_HEADERS` in `buf`.

rocallahan commented on src/Wine.cc:
```diff
@@ -0,0 +1,175 @@
+          struct IMAGE_SECTION_HEADER* sec = IMAGE_FIRST_SECTION(ntheader);
+          for (int i = 0; sec && i < std::max(ntheader->FileHeader.NumberOfSections, (uint16_t)1); ++i) {
+            struct IMAGE_SECTION_HEADER* seci = sec + i;
+            if ((char*)seci > buf + sizeof(buf)) break;
```
`break` on new line with {}

rocallahan replied:
Is it guaranteed that all the sections we care about are actually in the 1K `buf`?

bernhardu replied:
> Is it guaranteed that all the sections we care about are actually in the 1K `buf`?

I cannot guarantee, but this is also the size that [wine:gdbproxy.c](https://source.winehq.org/git/wine.git/blob/HEAD:/programs/winedbg/gdbproxy.c#l1518) uses.

rocallahan commented on src/Wine.cc:
```diff
@@ -0,0 +1,175 @@
+          for (auto it2 = mapping_list.begin(); name.empty() && it2 != mapping_list.end(); it2++) {
+            if (find(segments.begin(), segments.end(), it2->start()) != segments.end()) {
+              if (!it2->fsname().empty()) {
+                name = it2->fsname();
```
why not just drop the `name.empty()` check and `break` here?

rocallahan commented on src/Wine.cc:
```diff
@@ -0,0 +1,175 @@
+            continue;
+          }
+
+          // e.g. gdi32.so is prepared by wine to be found as PE image, but gdb does not like it.
```
Can you explain this more?

rocallahan commented on src/Wine.cc:
```diff
@@ -0,0 +1,175 @@
+  for (auto m = mapping_list.begin(); m != mapping_list.end(); ) {
+    bool del = false;
+    for (auto lib : libraries) {
+      if (m->fsname() == lib.first) del = true;
```
These if bodies need to be on their own line with {}

rocallahan commented on src/Wine.cc:
```diff
@@ -0,0 +1,175 @@
+        libraries[name].has_exec = true;
+      }
+    }
+  }
```
What if the same library is mapped twice for some reason?

rocallahan replied:
If some code creates an executable memory mapping that's not a file (e.g. JITted code), that will show up in `libraries` I guess. What will gdb do then?

bernhardu replied:
> What if the same library is mapped twice for some reason?

Current version would send addresses from both within one library to gdb. I guess this is one such a "fragility". Have to investigate.

bernhardu replied:
> If some code creates an executable memory mapping that's not a file (e.g. JITted code), that will show up in `libraries` I guess. What will gdb do then?

Currently this will just show up if there is a filename in the mapping present.

rocallahan commented on src/ReplayCommand.cc:
```diff
@@ -69,7 +69,10 @@ ReplayCommand ReplayCommand::singleton(
+    "                             be slower, but be able to tolerate missing files\n"
+    "  --target-wine              Tell gdb to use a library list supplied by rr.\n"
+    "                             This requires to jump after wine finished process\n"
+    "                             creation by adding option -g.\n");
```
Why do users have to use `-g` like this?

Could we automatically detect WINE instead?

bernhardu replied:
> Why do users have to use `-g` like this?

I did not get past this "Program stopped" message.

```
benutzer@debian:~/wine/drive_c/REVOLT$ rr replay wine-1
...
(rr) cont
Continuing.

Program stopped.
0x70000002 in syscall_traced ()
(rr) bt
#0  0x70000002 in syscall_traced ()
#1  0xb7f7e183 in _raw_syscall () at /home/bernhard/data/entwicklung/2021/rr/2021-04-09/rr/src/preload/raw_syscall.S:34
#2  0xb7f7c194 in traced_raw_syscall (call=0x681fffd8) at /home/bernhard/data/entwicklung/2021/rr/2021-04-09/rr/src/preload/syscallbuf.c:278
#3  syscall_hook_internal (call=0x681fffd8) at /home/bernhard/data/entwicklung/2021/rr/2021-04-09/rr/src/preload/syscallbuf.c:3360
#4  syscall_hook (call=<optimized out>) at /home/bernhard/data/entwicklung/2021/rr/2021-04-09/rr/src/preload/syscallbuf.c:3394
#5  0xb7f793e4 in _syscall_hook_trampoline () at /home/bernhard/data/entwicklung/2021/rr/2021-04-09/rr/src/preload/syscall_hook.S:139
#6  0xb7f7944d in _syscall_hook_trampoline_90_90_90 () at /home/bernhard/data/entwicklung/2021/rr/2021-04-09/rr/src/preload/syscall_hook.S:223
#7  0x6ffff005 in __kernel_vsyscall ()
#8  0xb7e4fc79 in execve () at ../sysdeps/unix/syscall-template.S:120
#9  0x00417f51 in ?? ()
#10 0x004184a9 in ?? ()
#11 0x00417edb in ?? ()
#12 0x00417270 in ?? ()
#13 0x00417963 in ?? ()
#14 0x00416650 in ?? ()
#15 0x0041e262 in ?? ()
#16 0x0041478f in ?? ()
#17 0xb7da3e46 in __libc_start_main (main=0x414590, argc=3, argv=0xbfc8b0d4, init=0x427b20, fini=0x427b80, rtld_fini=0xb7fa0230 <_dl_fini>, stack_end=0xbfc8b0cc) at ../csu/libc-start.c:308
#18 0x00414811 in ?? ()
(rr) when 
Current event: 1050
(rr) info inferior
  Num  Description       Connection                         Executable        
* 1    process 2011      1 (extended-remote 127.0.0.1:2235) /home/benutzer/.local/share/rr/wine-1/mmap_pack_213_dash 
(rr) kill
Kill the program being debugged? (y or n) y
[Inferior 1 (process 2011) killed]
(rr) q
benutzer@debian:~/wine/drive_c/REVOLT$ rr replay -g 1060 wine-1
...
Remote debugging using 127.0.0.1:2250

--------------------------------------------------
 ---> Reached target process 2011 at event 1061.
--------------------------------------------------
Reading symbols from /lib/ld-linux.so.2...
Reading symbols from /usr/lib/debug/.build-id/77/17f60b9dcc4447c397f1d52e9845e938cdca62.debug...
BFD: warning: system-supplied DSO at 0x6fffd000 has a section extending past end of file
0xb7fd8117 in get_extended_indices (cpu_features=<optimized out>) at ../sysdeps/x86/cpu-features.c:49
49      ../sysdeps/x86/cpu-features.c: Datei oder Verzeichnis nicht gefunden.
(rr) info inferior
  Num  Description       Connection                         Executable        
* 1    process 2011      1 (extended-remote 127.0.0.1:2250) /home/benutzer/.local/share/rr/wine-1/mmap_pack_165_wine 
(rr)
```

rocallahan commented:
One thing to think about here is how we're going to test this. Any ideas?

Keno commented:
Is there a reason that GDB doesn't just walk the dynamic linker data structures here like it does on Linux? That's how I used to do it for my rr frontend, which worked pretty well usually. Having all this logic in the debug server seems quite fragile.

bernhardu commented:
I am sorry for not replying earlier to the comments from Febrary, time is running too fast ...

> Is there a reason that GDB doesn't just walk the dynamic linker data structures here like it does on Linux? That's how I used to do it for my rr frontend, which worked pretty well usually. Having all this logic in the debug server seems quite fragile.

I am doing this for the PE images that wine loads. I am currently not aware that the linux dynamic linker gets notified of these dlls. But by controlling the PE libraries I supply to gdb, I was forced to supply the ELF mappings too. And doing it like in these patches worked enough for my tests at least.

rocallahan commented:
I guess the question Keno is asking is: shouldn't PE walking support be added to gdb instead of rr?

bernhardu commented:
> One thing to think about here is how we're going to test this. Any ideas?

A test that dlopens a .so and breaks to some function from it should be possible.
To generate and load a PE library with least effort a dependency to mingw and wine would be needed, but I guess is not wanted.
It might be possible to generate a fake dll in a test and somehow map this into memory to reach the PE code paths.

bernhardu commented:
> > > Is there a reason that GDB doesn't just walk the dynamic linker data structures here like it does on Linux? That's how I used to do it for my rr frontend, which worked pretty well usually. Having all this logic in the debug server seems quite fragile.
> > 
> > I am doing this for the PE images that wine loads. I am currently not aware that the linux dynamic linker gets notified of these dlls. But by controlling the PE libraries I supply to gdb, I was forced to supply the ELF mappings too. And doing it like in these patches worked enough for my tests at least.

> I guess the question Keno is asking is: shouldn't PE walking support be added to gdb instead of rr?

This seems to be a question that applies to wine itself too, why they have not changed gdb but instead doing their own debugger.

In my opinion from a gdb point of view, it would need to use wine infrastructure to achieve this.
From a wine point of view they could just use a gdb for windows (and I think winedbg without --gdb is more or less this).
Wouldn't this ask for a rr port to windows? :wink: 

I am sorry, but I still cannot say what it would really take for gdb and rr to support proper PE walking support.
Therefore I used this simple, maybe fragile, approach to just look at the memory mappings, and avoiding dependencies.

rocallahan commented:
I guess another question is: given that WINE chose to make their own debugger, have you considered integrating winedbg itself with rr?

Because if the WINE project is focused on using winedbg to debug WINE applications, it's probably not a good idea to invest in using gdb to debug WINE applications.

bernhardu committed:
Add simple test for --target-wine.

bernhardu commented:
> I guess another question is: given that WINE chose to make their own debugger, have you considered integrating winedbg itself with rr?
> 
> Because if the WINE project is focused on using winedbg to debug WINE applications, it's probably not a good idea to invest in using gdb to debug WINE applications.

winedbg has two modes of operation. In the first one winedbg does all itself, but has very simple user interface.
The other starts gdb as a proxy very similar to what normal rr does. So wine does use gdb too.
The changes in this patches are for a big part to replicate the setup of gdb started by rr to be simliar to one started by winedbg.

bernhardu commented:
I made these changes in this iteration:
- Rewrote the PE describing structures.
- Changed formatting of if statements.
- Using reinterpret_cast instead of c-style casts.
- Avoided creation of pointers before checking if they will point to valid addresses.
- Don't announce exec-file, executable gets delivered as library with --target-wine for now.
- Add simple test.
- Rebased to latest upstream head.


Still open:
- Better way to enumerate PE- and ELF libraries? Currently retrieved from memory mappings.
- Better way to convince gdb to use qXfer:libraries mode?
- Make unix function support libraries supported. (E.g. gdi32.so on top of gdi32.dll)
- Library loaded twice?
- Is 1K buffer enough for all sections?
- Automatically detect wine? (Currently not even attempted on purpose to minimize influence normal rr mode.)
- Improve test.
- Integrate winedbg directly within rr / differently?

bernhardu commented:
Just for information:
At the wine mailing list [a gdb script](https://www.winehq.org/pipermail/wine-devel/2021-June/thread.html#189134) was presented, that is doing add-symbol-file's based on the content of /proc/pid/maps, so most of what the purpose of this change initially was.

bernhardu commented:
@GitMensch a short addition to my last message.
Today I found another point one has to take care differently for wine.
Wine removes some environment variables, e.g. `QT_NO_CPU_FEATURE`,
therefore a windows Qt application executes some problematic instructions.
(This is removal done in wine in `is_special_env_var`.)
That way this recording is stopping with a `Assertion 'ticks_now == trace_ticks' failed to hold. ticks mismatch for '...'; expected 29070847, got 29070851`.
So either wine has to be modified not to clear this variable,
or start recording by just executing cmd, set the variable yourself and then execute the application.
Happy recording of wine :slightly_smiling_face:

Keno commented:
> Is there a reason that GDB doesn't just walk the dynamic linker data structures here like it does on Linux? That's how I used to do it for my rr frontend, which worked pretty well usually. Having all this logic in the debug server seems quite fragile.

I took the liberty of implementing this, since I needed to do some wine debugging and my own frontend is a bit bitrotted. See here: https://github.com/JuliaComputing/gdb-solib-wine. Seems to work quite well.

bernhardu commented:
Awesome :smiley:. I made a little test and it works beautifully. I did run run rr, wine and gdb from thier build directories. Therefore I put to rr the `--debugger` parameter and I guess therefore I needed to give the file command an absolute path (giving the mmap file from the recording worked too) and `set solib-search-path`. Thanks for your work.

I guess this makes this PR obsolete, so I we are finished here and I am going to close it.

Keno commented:
> Therefore I put to rr the --debugger parameter and I guess therefore I needed to give the file command an absolute path (giving the mmap file from the recording worked too) and set solib-search-path. Thanks for your work.

Yes, I'm planning to improve that, there's a few todos still.
